index.ts

```diff
@@ -104,541 +119,285 @@ class Rabbitr extends EventEmitter {
+    // start by opening an AMQP connection
+    this.connection = await amqplib.connect(this.opts.url, this.opts.connectionOpts);
+
+    // make sure to close the connection if the process terminates
+    process.once('SIGINT', this.shutdown);
+    process.once('SIGTERM', this.shutdown);
+    this.connection.on('close', () => {
+      process.removeListener('SIGINT', this.shutdown);
+      process.removeListener('SIGTERM', this.shutdown);
+      if (!this.isShuttingDown) {
+        throw new Error('Disconnected from RabbitMQ');
+      }
+      this.log(`connection closed`);
+    });
 
-    return name;
+    // now setup all the channels necessary for standard operation
+    const channels = await Promise.all([
+      this.connection.createChannel(),
+      this.connection.createChannel(),
+    ]);
+    this._timerChannel = channels[0];
+    this._publishChannel = channels[0];
+    this._backoffChannel = channels[0];
+    this._cachedChannel = channels[0];
+    this._rpcReturnChannel = channels[1];
+    this._openChannels = this._openChannels.concat(channels);
+
+    // mark the connection as open and perform any 'setup' method if one is set
+    this.connected = true;
+    await maybeFromCallback<void>(this.opts.setup || (() => Promise.resolve()));
+
+    // mark this instance as 'ready' and end the promise
+    this.ready = true;
+    return this.connection;
   }
 
-  private destroyPromise: Bluebird<void>;
   /** method to destroy anything for this instance of rabbitr */
-  destroy(cb?: Rabbitr.ErrorCallback): Bluebird<void> {
-    if (!this.destroyPromise) {
-      this.log(`${red('destroying')}`);
-      this.destroyPromise = Bluebird.each(this._openChannels, channel =>
-        Bluebird.fromCallback(channel.close.bind(channel))
-          .then(
-            () => this.log('channel closed'),
-            err => {
-              // istanbul ignore next
-              this.log('Error while closing connection', err);
-              throw err;
-            }
-          )
-      ).then(() => {
-        return Bluebird.fromCallback(callback =>
-          this.connection.close(callback)
-        ).then(
-          () => {
-            this.log('connection closed');
-            this.removeAllListeners();
-            this.connectionPromise = null;
-          },
-          err => {
-            // istanbul ignore next
-            this.log('Error while closing connection', err);
-            throw err;
-          }
-        );
-      });
+  async destroy() {
+    if(!this.connected) {
+      this.log(`${red('unable to destroy this instance as it is not connected')}`);
+      return;
     }
-    return this.destroyPromise.asCallback(cb);
+
+    // mark this instance as no longer valid
+    this.log(`${red('destroying')}`);
+    this.ready = false;
+    this.connected = false;
+
+    // close all the channels that are open
+    await Promise.all(this._openChannels.map(async (channel) => {
+      await channel.close.bind(channel);
```

Comment:
`.bind()` produces new function. What's the point of it here? Didn't you want to do `.call(channel)` instead? BTW what's the point of setting context explicitly ? Hasn't .close got channel instance as "this" ? Is it assigned to 'channel' after instantiation ?

Reply:
Looks like it was originally passed to `.map` like this: `.map(channel.close.bind(channel))` and was incorrectly copy-pasted.